rfcs/0062-content-addressed-paths.md

```diff
@@ -0,0 +1,310 @@
+- known to be deterministic (for caching reasons, see [caching]).
+
+That way we don't have to worry about the fact that hash-rewriting is only an
+approximation nor by the semantics of the distribution of non-deterministic
+paths.
+
+We also leave the option to lift these restrictions later.
+
+This RFC already has a (somewhat working) POC at
+<https://github.com/NixOS/nix/pull/3262>.
+
+# Motivation
+
+[motivation]: #motivation
+
+Having a content-adressed store with Nix (aka the "Intensional store") is a
+long-time dream of the community − a design for that was already taking a whole
+chapter in [Eelco's PHD thesis][nixphd].
+
+This was never done because it represents a quite big change in Nix's model,
+with some non-totally-solved implications (regarding the trust model in
+particular).
+Even without going all the way down to a fully intensional model, we can
+make specific paths content-adressed, which can give some important benefits of
+the intensional store at a much lower price. In particular, setting some
+critical derivations as content-adressed can lead to some substancial build
+cutoffs.
+
+# Detailed design
+
+[design]: #detailed-design
+
+The gist of the design is that:
+
+- Some derivations can be marked as content-adressed (ca), in which case each
+  one of their output will be moved to a path `ca` determined only by its
+  content after the build
+- We introduce the notion of a `resolved derivation` which is a derivation that
+  doesn't refer to any other derivation but only to concrete store paths.
+  To prevent ambiguities, we might speak of a `symbolic derivation` to
+  designate a derivation that's not necessarily resolved.
+  We also define a `resolving` function that given a symbolic derivation
+  returns a new resolved derivation with the same semantics.
+- When asked to build a derivation, Nix will first resolve it, build the
+  resolved derivation and link back the symbolic one to the out path of the
+  resolved one.
+
+## Nix-build process
+
+### Output mappings
+
+For each output `output` of a derivation `drv`, we define
+
+- its output id **DrvOutputId(drv, output)** as the tuple `(hash(drv), output, truster)`, where `truster` is a reserved field for future use and currently always set to `"world"`.
+  This id uniquely identifies the output.
+  We textually represent this as `hash(drv)!output[@truster]`.
+- its concrete path **PathOf(outputId)** as the path on which the output will be stored on disk.
+
+> Unresolved: should we already include the `truster` field in `DrvOutputId`
+> even if it's not used atm? What would be the cost of adding it later?
+
+In a dependency-addressed-only world, the concrete path for a derivation output was a pure function of this output's id that could be computed at eval-time. However this won't be the case anymore once we allow content-addressed derivations, so we now need to store the results the `PathOf` function in the Nix database as a new table:
```

Comment:
Is there a difference between `dependency-addressed` and `content-addressed`? If no, we should stick with one term.

Reply:
Ah, `dependency-addressed` is the term for the status of nix as-is? If you want to introduce this terminology here, you have to also define it somehow.